In [2]:
import math

### Factors affecting derivetive prices
<b>Variables</b> are underlying asset value $S$ and time to expiry $t$.<br/>
<b>Parameters</b> are interest rate, strike price, dividends, <b>volatility</b>.<br/>
<br/>
<b>Volatility</b> is annualized standard deviation of asset returns.<br/>

#### Put Call Parity
$$\max(S(T) - E, 0) - \max(E-S(T), 0) = S(T) - E$$
at expiration time<br/>
or
$$C - P = S - E e^{-r(T-t)}$$

#### Binary Put-Call Parity
$Binary Call + Binary Put = e^{-r(T-t)}$

## Options strategies

#### Bull Spread
is portfolio of long call option with strike $E_1$ and short call option with strike $E_2$, where $E_2 > E_1$ <br/>
Payoff function is 
$$\frac{1}{E_2 - E_1} (\max{(S-E_1, 0)} - \max{(S-E_2, 0)})$$

#### Bear Spread
is portfolio of long put with string $E_1$ and short put with strike $E_2$, where $E_2 > E_1$
<br/>
<br/>
<i><font color="red">Definition:</font></i> Strategy involving options of the same type (i.e. calls or puts) is called <b>spread</b>.

#### Straddle
is long call and long put with the same strike (bet on rising volatility)

#### Strangle
is long call and long put with different strices ($E_{call} > E_{put}$)

#### Risk Reversal
<b>Long call</b> with strike above corrent spot + <b>short put</b> with strike below current spot.

#### Butterfly spread
Long calls with strikes $E_1$ and $E_3$ and short two calls with strike $E_2$, where $E_1 < E_2 < E_3$

#### Calendar spread
is strategy involving options with different expiry dates.

-------------------------
##### <font color="orange">Exercise 1 [Done]</font>

(a)
\begin{equation}
\Lambda(S) = 
\begin{cases}
E , \quad & S<E \\
S, \quad & S \geq E
\end{cases}
\end{equation}

(b) $\Lambda(S) = |S-E|$ <br/>

(c) <br/>

\begin{equation}
\Lambda(S) = 
\begin{cases}
0, \quad & S <E_1 \\
S - E_1 , \quad & S \in [E_1, E_2] \\
E_2 - E_1 , \quad & S>E_2
\end{cases}
\end{equation}

(d) case $E_1 < E_2$
\begin{equation}
\Lambda(S) = 
\begin{cases}
E_2 - S, \quad & S < E_1 \\
E_2 - E_1, \quad & S \in [E_1, E_2) \\
S - E_1, \quad & S \geq E_2
\end{cases}
\end{equation}

case $E_1 = E_2 = E$
$$ |S-E| $$

case $E_1 > E_2$
\begin{equation}
\Lambda(S) = 
\begin{cases}
E_2 - S, \quad & S < E_2 \\
0, \quad & S \in [E_2, E_1) \\
S - E_1, \quad & S \geq E_1
\end{cases}
\end{equation}


(e) $E_1 < E < E_2$
\begin{equation}
\Lambda(S) = 
\begin{cases}
0, \quad & S < E_1 \\
S - E_1, \quad & S \in [E_1, E) \\
2E-S-E_1, \quad & S \in [E, E_2) \\
2E-E_1-E_2, \quad & S \geq E_2
\end{cases}
\end{equation}

##### <font color="orange">Exercise 2 [Done]</font>
<b>Payoff</b> diagram does not include premium paid for holding or received from writing. <br/> 
<b>Profit</b> diagram includes paid or received premiums.

\begin{equation}
\Lambda(S) = 
\begin{cases}
-S, \quad & S<E \\
-S + 2(S-E) = S-2E, \quad & S \geq E
\end{cases}
\end{equation}

##### <font color="orange">Exercise 3 [Done]</font>
$S = 60$<br/>
$E = 58$<br/>
$C = 3$<br/>
$P = 1.5$<br/>
$r = 5%$<br/>

In [4]:
60 - 58 * math.exp(-0.05/4)

2.720487571354873

There is an arbitrage opportunity
$C-P = 3 - 1.5 = 1.5 \neq 2.72 = S - E e^{-r(T-t)}$

##### <font color="orange">Exercise 4 [Done]</font>
11.91 - 2 * 4.52 + 1.03 = 3.9

In [8]:
11.91 - 2 * 4.52 + 1.03 

3.9000000000000012

##### Exercise 5 